README.md
# cap-consumer

[![Build Status](https://app.travis-ci.com/lykmapipo/cap-consumer.svg?branch=master)](https://app.travis-ci.com/lykmapipo/cap-consumer)
[![Dependencies Status](https://david-dm.org/lykmapipo/cap-consumer.svg)](https://david-dm.org/lykmapipo/cap-consumer)
[![Coverage Status](https://coveralls.io/repos/github/lykmapipo/cap-consumer/badge.svg?branch=master)](https://coveralls.io/github/lykmapipo/cap-consumer?branch=master)
[![GitHub License](https://img.shields.io/github/license/lykmapipo/cap-consumer)](https://github.com/lykmapipo/cap-consumer/blob/master/LICENSE) 

[![Commitizen Friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Code Style](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)
[![npm version](https://img.shields.io/npm/v/@lykmapipo/cap-consumer)](https://www.npmjs.com/package/@lykmapipo/cap-consumer)

Common Alert Protocol Consumer

## Requirements

- [NodeJS v12+](https://nodejs.org)
- [Npm v6+](https://www.npmjs.com/)

## Installation

```sh
npm install --save @lykmapipo/cap-consumer
```

## Usage

```js
import { fetchAlerts } from '@lykmapipo/cap-consumer';

const optns = { url: 'https://cap-sources.s3.amazonaws.com/tz-tma-en/rss.xml' };
fetchAlerts(optns)
  .then(alerts => { ... }) // => { channel: { ... }, items: [ ... ] }
  .catch(error => { ... });
```

## Test

- Clone this repository

- Install all dependencies

```sh
npm install
```

- Then run test

```sh
npm test
```

## References
- [RSS 2.0 Specification](https://cyber.harvard.edu/rss/rss.html)
- [Common_Alerting_Protocol](https://en.wikipedia.org/wiki/Common_Alerting_Protocol)
- [Common Alerting Protocol Version 1.2](http://docs.oasis-open.org/emergency/cap/v1.2/CAP-v1.2-os.html)

## Contribute

It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## LICENSE

MIT License

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 0.2.9 (2021-09-06)

##### Chores

* **deps:**  force latest version & audit fix ([13a55e01](https://github.com/lykmapipo/cap-consumer/commit/13a55e013339f2de8aaab59330c08186e37910d3))
*  upgrade package scripts ([e905df19](https://github.com/lykmapipo/cap-consumer/commit/e905df197943c55f86e66fafd9a6da83eadafdb6))

##### Documentation Changes

* **readme:**  update travis badge source ([2bc50c5d](https://github.com/lykmapipo/cap-consumer/commit/2bc50c5dc08137a46cf50a6b9e85e508ec88de81))
* **community:**  add security guide ([facea790](https://github.com/lykmapipo/cap-consumer/commit/facea7901a374d78c25a097462359e1bea7cccc7))

##### Refactors

* **deps:**  use latest version ([6083915a](https://github.com/lykmapipo/cap-consumer/commit/6083915ab0e010ffc3ce543c7b15cb44c83708f5))

#### 0.2.8 (2021-02-10)

##### Chores

* **deps:**  force latest version & audit fix ([e377b41e](https://github.com/lykmapipo/cap-consumer/commit/e377b41e6bd99fd87d1aabe71bbb9381e452374e))

#### 0.2.7 (2021-01-06)

##### Chores

* **deps:**  force latest version & audit fix ([a22149c7](https://github.com/lykmapipo/cap-consumer/commit/a22149c7eb9d7bf3e8cf88f475e03ea33a166bd6))

#### 0.2.6 (2020-12-26)

##### Chores

* **deps:**  force latest version & audit fix ([3e3c9b8e](https://github.com/lykmapipo/cap-consumer/commit/3e3c9b8e1f67a1270ef841a8f25d47f12bcbbf83))

#### 0.2.5 (2020-12-17)

##### Chores

* **deps:**  force latest version & audit fix ([5daf8edc](https://github.com/lykmapipo/cap-consumer/commit/5daf8edc558bbf059f185706fbe2012b1db5036d))

#### 0.2.4 (2020-12-06)

##### Chores

* **deps:**  force latest version & audit fix ([7225b3a0](https://github.com/lykmapipo/cap-consumer/commit/7225b3a0f9993357e6b3f52fd6a464cf0c60fa35))

#### 0.2.3 (2020-10-05)

##### Chores

* **deps:**  force latest version & audit fix ([cbf14e1f](https://github.com/lykmapipo/cap-consumer/commit/cbf14e1fd67aca3d763ae010ac6b2ac4892a5d51))

#### 0.2.2 (2020-09-23)

#### 0.2.1 (2020-09-10)

##### Chores

* **deps:**  force latest version & audit fix ([e94f8706](https://github.com/lykmapipo/cap-consumer/commit/e94f870629079ec331ec4a22cad5f5ae7d9b98cc))

#### 0.2.0 (2020-09-06)

##### Refactors

*  use cap commons ([49d7dcae](https://github.com/lykmapipo/cap-consumer/commit/49d7dcaead955eceb518c22128c0b41425e2cfc6))

#### 0.1.25 (2020-09-05)

##### Chores

* **deps:**  force latest version & audit fix ([8620099e](https://github.com/lykmapipo/cap-consumer/commit/8620099e5cd67d21c558669062a74727a4b238bb))

##### Tests

* **fixtures:**  update images ([fc382ec8](https://github.com/lykmapipo/cap-consumer/commit/fc382ec86d839c081afb4610db4f392b6a3a87bd))
*  fix fetch alerts n-times calls ([8b23800f](https://github.com/lykmapipo/cap-consumer/commit/8b23800f41452f3b6c029043a9e67c1b2aab30f5))

#### 0.1.24 (2020-09-02)

##### Chores

* **deps:**  force latest version & audit fix ([5bf73fd3](https://github.com/lykmapipo/cap-consumer/commit/5bf73fd3243082d86917cfbb892f7b764ad802b2))

#### 0.1.23 (2020-08-19)

##### Chores

* **deps:**  force latest version & audit fix ([e6e07f39](https://github.com/lykmapipo/cap-consumer/commit/e6e07f393738c80fb051b6ce944d85f4b72fb005))

#### 0.1.22 (2020-08-04)

##### Chores

* **deps:**  force latest version & audit fix ([ede32938](https://github.com/lykmapipo/cap-consumer/commit/ede32938f2d7a444c0e104eace75b5331dadcee7))

#### 0.1.21 (2020-07-27)

##### Chores

* **deps:**  force latest version & audit fix ([6c2929fb](https://github.com/lykmapipo/cap-consumer/commit/6c2929fb63784bee77d59755010e0cce9c30c59b))

#### 0.1.20 (2020-07-20)

##### Chores

* **deps:**  force latest version & audit fix ([bce211f8](https://github.com/lykmapipo/cap-consumer/commit/bce211f879aa40ea1c2879f2c70d432de6916e0b))

#### 0.1.19 (2020-07-12)

##### Chores

* **deps:**  force latest version & audit fix ([3a2eceac](https://github.com/lykmapipo/cap-consumer/commit/3a2eceac43e33adee11443ac9b0ace71b5501ea1))

#### 0.1.18 (2020-07-08)

##### Chores

* **deps:**  force latest version & audit fix ([07c36a96](https://github.com/lykmapipo/cap-consumer/commit/07c36a9609664e6d4a8fc83543f372b14325f7c5))

#### 0.1.17 (2020-07-01)

##### Chores

* **deps:**  force latest version & audit fix ([9c6aaf83](https://github.com/lykmapipo/cap-consumer/commit/9c6aaf83e6dd65cb65318a58dce9abeb33e44321))

#### 0.1.16 (2020-06-25)

##### Chores

* **deps:**  force latest version & audit fix ([ab822128](https://github.com/lykmapipo/cap-consumer/commit/ab82212847d6c8b867f4b3832e8d7de2bd732650))

#### 0.1.15 (2020-06-16)

##### Chores

* **deps:**  force latest version & audit fix ([4c7255e3](https://github.com/lykmapipo/cap-consumer/commit/4c7255e37fd88a2cd2cb28b781a07fce5377804c))

#### 0.1.14 (2020-06-07)

#### 0.1.13 (2020-05-29)

##### Chores

* **deps:**  force latest version & audit fix ([1b1e4486](https://github.com/lykmapipo/cap-consumer/commit/1b1e4486d3cd0e5fb8b6fe001494fc0911933005))

#### 0.1.12 (2020-05-14)

##### Chores

* **deps:**  force latest version & audit fix ([8676da86](https://github.com/lykmapipo/cap-consumer/commit/8676da863e92f76aa23aec84b3cc3d123edddaf1))

#### 0.1.11 (2020-05-09)

##### Chores

* **deps:**  force latest version & audit fix ([7ee19f26](https://github.com/lykmapipo/cap-consumer/commit/7ee19f261f09fb24a95e428046e52a404d8f27c3))

#### 0.1.10 (2020-05-01)

##### Chores

* **deps:**  force latest version & audit fix ([90ebe138](https://github.com/lykmapipo/cap-consumer/commit/90ebe13878ed62dd1635976966d77b2b7377e9f8))

#### 0.1.9 (2020-04-25)

##### Chores

* **deps:**  force latest version & audit fix ([de664a4a](https://github.com/lykmapipo/cap-consumer/commit/de664a4a881a104028a617146b78f0bfbfaa5355))

#### 0.1.8 (2020-04-15)

##### Chores

* **deps:**  force latest version & audit fix ([bc56f6cc](https://github.com/lykmapipo/cap-consumer/commit/bc56f6cc3217b8440ec1895939fb838b72f5d5db))

#### 0.1.7 (2020-03-30)

##### Chores


test/index.spec.js
import { createReadStream, readFileSync } from 'fs';
import { expect, nock } from '@lykmapipo/test-helpers';

import {
  parseAlert,
  parseFeed,
  fetchAlert,
  fetchFeed,
  fetchAlerts,
} from '../src';

const BASE_URL = 'https://cap-sources.s3.amazonaws.com/tz-tma-en';

describe('consumer', () => {
  beforeEach(() => {
    nock.cleanAll();
  });

  it('should parse alert xml to json', (done) => {
    const xml = readFileSync(`${__dirname}/fixtures/alert.xml`, 'utf-8');
    parseAlert(xml)
      .then((alert) => {
        expect(alert).to.exist;
        expect(alert.identifier).to.exist;
        expect(alert.sender).to.exist;
        expect(alert.sent).to.exist;
        expect(alert.status).to.exist;
        expect(alert.msgType).to.exist;
        expect(alert.scope).to.exist;
        expect(alert.references).to.exist;
        expect(alert.info).to.exist;
        expect(alert.info.language).to.exist;
        expect(alert.info.category).to.exist;
        expect(alert.info.event).to.exist;
        expect(alert.info.responseType).to.exist;
        expect(alert.info.urgency).to.exist;
        expect(alert.info.severity).to.exist;
        expect(alert.info.certainty).to.exist;
        expect(alert.info.onset).to.exist;
        expect(alert.info.expires).to.exist;
        expect(alert.info.senderName).to.exist;
        expect(alert.info.headline).to.exist;
        expect(alert.info.description).to.exist;
        expect(alert.info.instruction).to.exist;
        expect(alert.info.web).to.exist;
        expect(alert.info.contact).to.exist;
        expect(alert.info.area).to.exist;
        expect(alert.info.area.areaDesc).to.exist;
        expect(alert.info.area.geometry).to.exist;
        expect(alert.info.area.centroid).to.exist;
        expect(alert.hash).to.exist;
        done(null, alert);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  it('should parse alerts feed', (done) => {
    const source = createReadStream(`${__dirname}/fixtures/feed.xml`);
    parseFeed(source)
      .then((feed) => {
        expect(feed).to.exist.and.be.an('object');
        expect(feed.channel).to.exist.and.be.an('object');
        expect(feed.items).to.exist.and.be.an('array');
        done(null, feed);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  it('should fetch alert', (done) => {
    nock(BASE_URL)
      .get('/alert.xml')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/alert.xml`);
      });

    fetchAlert({ url: `${BASE_URL}/alert.xml` })
      .then((alert) => {
        expect(alert).to.exist;
        expect(alert.identifier).to.exist;
        expect(alert.sender).to.exist;
        expect(alert.sent).to.exist;
        expect(alert.status).to.exist;
        expect(alert.msgType).to.exist;
        expect(alert.scope).to.exist;
        expect(alert.references).to.exist;
        expect(alert.info).to.exist;
        expect(alert.info.language).to.exist;
        expect(alert.info.category).to.exist;
        expect(alert.info.event).to.exist;
        expect(alert.info.responseType).to.exist;
        expect(alert.info.urgency).to.exist;
        expect(alert.info.severity).to.exist;
        expect(alert.info.certainty).to.exist;
        expect(alert.info.onset).to.exist;
        expect(alert.info.expires).to.exist;
        expect(alert.info.senderName).to.exist;
        expect(alert.info.headline).to.exist;
        expect(alert.info.description).to.exist;
        expect(alert.info.instruction).to.exist;
        expect(alert.info.web).to.exist;
        expect(alert.info.contact).to.exist;
        expect(alert.info.area).to.exist;
        expect(alert.info.area.areaDesc).to.exist;
        expect(alert.info.area.geometry).to.exist;
        expect(alert.info.area.centroid).to.exist;
        expect(alert.hash).to.exist;
        done(null, alert);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  it('should fetch alerts feed', (done) => {
    nock(BASE_URL)
      .get('/feed.xml')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/feed.xml`);
      });

    fetchFeed({ url: `${BASE_URL}/feed.xml` })
      .then((feed) => {
        expect(feed).to.exist.and.be.an('object');
        expect(feed.channel).to.exist.and.be.an('object');
        expect(feed.items).to.exist.and.be.an('array');
        done(null, feed);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  it('should fetch alerts', (done) => {
    nock(BASE_URL)
      .get('/feed.xml')
      .query(true)
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/feed.xml`);
      });

    nock(BASE_URL)
      .get('/alert.xml') // match alerts url
      .query(true)
      .times(2) // we have only 2 items
      .reply(200, function onReply() {
        expect(this.req.headers).to.exist;
        return createReadStream(`${__dirname}/fixtures/alert.xml`);
      });

    fetchAlerts({ url: `${BASE_URL}/feed.xml` })
      .then((alerts) => {
        expect(alerts).to.exist.and.be.an('object');
        expect(alerts.channel).to.exist.and.be.an('object');
        expect(alerts.items).to.exist.and.be.an('array');
        expect(alerts.items).to.have.length(2);

        const alert = alerts.items[0];
        expect(alert).to.exist;
        expect(alert.identifier).to.exist;
        expect(alert.sender).to.exist;
        expect(alert.sent).to.exist;
        expect(alert.status).to.exist;
        expect(alert.msgType).to.exist;
        expect(alert.scope).to.exist;
        expect(alert.references).to.exist;
        expect(alert.info).to.exist;
        expect(alert.info.language).to.exist;
        expect(alert.info.category).to.exist;
        expect(alert.info.event).to.exist;
        expect(alert.info.responseType).to.exist;
        expect(alert.info.urgency).to.exist;
        expect(alert.info.severity).to.exist;
        expect(alert.info.certainty).to.exist;
        expect(alert.info.onset).to.exist;
        expect(alert.info.expires).to.exist;
        expect(alert.info.senderName).to.exist;
        expect(alert.info.headline).to.exist;
        expect(alert.info.description).to.exist;
        expect(alert.info.instruction).to.exist;
        expect(alert.info.web).to.exist;
        expect(alert.info.contact).to.exist;
        expect(alert.info.area).to.exist;
        expect(alert.info.area.areaDesc).to.exist;
        expect(alert.info.area.geometry).to.exist;
        expect(alert.info.area.centroid).to.exist;
        expect(alert.hash).to.exist;

        done(null, alerts);
      })
      .catch((error) => {
        expect(error).to.not.exist;
        done(error);
      });
  });

  afterEach(() => {
    nock.cleanAll();
  });
});


test/utils.spec.js
import { expect, faker } from '@lykmapipo/test-helpers';
import { processors } from 'xml2js';

import {
  DEFAULT_REQUEST_HEADERS,
  XML_PARSE_OPTIONS,
  normalize,
} from '../src/utils';

describe('utils', () => {
  it('should provide default request headers', () => {
    expect(DEFAULT_REQUEST_HEADERS).to.be.eql({
      accept: 'application/xhtml+xml',
      'content-type': 'application/xhtml+xml',
      'user-agent':
        'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36',
    });
  });

  it('should provide xml parse options', () => {
    expect(XML_PARSE_OPTIONS).to.be.eql({
      trim: true,
      explicitArray: false,
      explicitRoot: false,
      tagNameProcessors: [processors.stripPrefix],
    });
  });

  it('should normalize object value', () => {
    const obj = {
      uid: faker.datatype.uuid(),
      rss: faker.datatype.uuid(),
      atom: faker.datatype.uuid(),
      '#': faker.datatype.uuid(),
      '@': faker.datatype.uuid(),
      $: faker.datatype.uuid(),
    };
    const normalized = normalize(obj);
    expect(normalized).to.be.eql({ uid: obj.uid });
    expect(normalized).to.not.have.a.property('rss');
    expect(normalized).to.not.have.a.property('atom');
    expect(normalized).to.not.have.a.property('#');
    expect(normalized).to.not.have.a.property('@');
    expect(normalized).to.not.have.a.property('$');

    expect(normalize()).to.be.eql({});
  });
});


src/utils.js
import { omitBy, startsWith } from 'lodash';
import { mergeObjects } from '@lykmapipo/common';
import { processors } from 'xml2js';

/**
 * @constant DEFAULT_REQUEST_HEADERS
 * @name DEFAULT_REQUEST_HEADERS
 * @description Default http request headers
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @private
 * @ignore
 */
export const DEFAULT_REQUEST_HEADERS = {
  accept: 'application/xhtml+xml',
  'content-type': 'application/xhtml+xml',
  'user-agent':
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36',
};

/**
 * @constant XML_PARSE_OPTIONS
 * @name XML_PARSE_OPTIONS
 * @description Default xml parse options
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @private
 * @ignore
 */
export const XML_PARSE_OPTIONS = {
  trim: true,
  explicitArray: false,
  explicitRoot: false,
  tagNameProcessors: [processors.stripPrefix],
};

/**
 * @function normalize
 * @name normalize
 * @description Normalize object to remove unused properties
 * @param {object} obj valid object to normalize.
 * @returns {object} normalize object
 * @author lally elias <lallyelias87@mail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * normalize(obj);
 * // => { ... }
 */
export const normalize = (obj = {}) => {
  // remove unused properties
  const normalized = omitBy(obj, (value, key) => {
    return (
      startsWith(key, 'rss') ||
      startsWith(key, 'atom') ||
      startsWith(key, '#') ||
      startsWith(key, '@') ||
      startsWith(key, '$') ||
      startsWith(key, 'Signature')
    );
  });
  // return
  return mergeObjects(normalized);
};
